### PR TITLE
Optimize ContentPath#pathAsText

### PR DESCRIPTION
**Pull request by dorukguner**
This pr optimizes calls to ContentPath#pathAsText through the following changes:
- When at the root we return the provided string directly
- When called on a sibling or child field the previously built path is maintained to prevent appending the same top-level fields again

The case in which fields that have been previously added to the path are removed is handled in ContentPath#remove by clearing the existing StringBuilder.

Tests for ContentPath#pathAsText have been added in a separate commit to the optimization to allow for easy validation that the tests are correct.

Addresses #94544

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java [hunk 1]
```diff
@@ -15,6 +15,7 @@ public final class ContentPath {
     private final StringBuilder sb;
 
     private int index = 0;
+    private int sbIndex = 0;
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java:18]: I think we can get rid of this additional index if we change the remove method so it resets the strin builder length according to the length of the element removed.

File: server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java [hunk 2]
```diff
@@ -44,6 +45,12 @@ private void expand() {
 
     public void remove() {
         path[--index] = null;
+
+        // Reset the StringBuilder if it includes a newly removed field
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java:49]: I think we can avoid a full reset here. We can determine the length of the removed path element before nulling it out, then reset the string builder length accordingly. Some special casing for index == 0 and the delimiters might be needed.

File: server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java [hunk 3]
```diff
@@ -55,12 +62,18 @@ public boolean isWithinLeafObject() {
...
-            sb.append(path[i]).append(DELIMITER);
+        // If index is 0 we know that we are at the root, so return the provided string directly
+        if (index == 0) {
+            return name;
         }
-        sb.append(name);
-        return sb.toString();
+
+        // Otherwise we preserve the previously built StringBuilder and append any necessary path fields
+        while (sbIndex < index) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java:71]: I think we can already keep the string builder current while we add / remove things and avoid the need to have a second index here

File: server/src/test/java/org/elasticsearch/index/mapper/ContentPathPathAsTextTests.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Collections;
+import java.util.List;
+
+public class ContentPathPathAsTextTests extends ESTestCase {
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/index/mapper/ContentPathPathAsTextTests.java:16]: Great tests, thanks for adding those, it makes code changes like these so much easier to check. 
Unfortunately there was another test class for ContentPath added with #98332 (ContentPathTests).
I would like to keep that one as the main test class. Would you mind to merge the test cases that are not covered to that unit test class? I think there might be some duplication now.

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @dorukguner, sorry for the delay in getting to this and thanks for picking it up.
I like the solution and the tests are great. Since we have tests now I would like to go a step further and optimize your solution even a bit more. I left a few suggestions for that, let me know what you think.

**Comment by dorukguner** [comment]
Hi @cbuescher, thanks for the feedback.

I've merged the extra test cases I had into the new main test class and I've also optimised ContentPath a little further.

I realised we didn't need both an array of strings for the path and a string builder. We can have just a string builder while keeping track of the indexes in which the delimiters were added, and using the add and remove methods to modify the string builder directly. This has removed the need for any loops and any extra indexes.

Please let me know what you think about the new optimisations, I'm open to any other suggestions you may have.

File: server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java [hunk 5]
```diff
@@ -8,42 +8,30 @@
...
+        // Store the location of the previous final delimiter onto the stack,
+        // which will be the index of the 2nd last delimiter after appending the new name
+        delimiterIndexes.add(sb.length() - 1);
+        sb.append(name).append(DELIMITER);
     }
-
     public void remove() {
-        path[--index] = null;
+        // Deletes the last node added to the stringbuilder by deleting from the 2nd last delimiter onwards
+        sb.delete(delimiterIndexes.pop() + 1, sb.length());
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java:34]: Maybe we can check the stack size and throw a nicer exception than the EmptyStackException, maybe something that mentions an empty contentPath.

**Review by cbuescher — COMMENTED** [review]
@dorukguner thanks a lot, I think the idea of just keeping one datastructure for the whole path and keeping around the pointers to the parts is great and the tests are also covering this nicely. I left two small comments but will prepare this PR for testing and merging while you address those.

**Comment by cbuescher** [comment]
Just running a check before the final changes to see if this breaks anything else.

@elasticsmachine test this please

**Comment by cbuescher** [comment]
@dorukguner could you also run ./gradlew :server:spotlessApply to apply our formatting rules to the modified classes?

**Review by cbuescher — APPROVED** [review]
Thanks for the update, LGTM. Will need to run full test suite before merging.

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by cbuescher** [comment]
@elasticmachine update branch

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by cbuescher** [comment]
@dorukguner huzzah, all passed. Thanks for the change and the iterations, merged to main branch now. I think this is a great change that makes using this class a lot less wasteful.

**Pull request merged**